internal/controllers/configuration/object_references.go

```diff
@@ -0,0 +1,180 @@
+package configuration
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	extv1beta1 "k8s.io/api/extensions/v1beta1"
+	netv1 "k8s.io/api/networking/v1"
+	netv1beta1 "k8s.io/api/networking/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/annotations"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/controllers/reference"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane"
+	kongv1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1"
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1beta1"
+)
+
+// updateReferredObjects updates reference records where the referrer is the object in parameter obj.
+// currently it only updates reference records to secrets, since we wanted to limit cache size of secrets:
+// https://github.com/Kong/kubernetes-ingress-controller/issues/2868
+func updateReferredObjects(
+	ctx context.Context, client client.Client, dataplaneClient *dataplane.KongClient, obj client.Object) error {
+	switch obj := obj.(type) {
+	// functions update***ReferredSecrets first list the secrets referred by object,
+	// then call UpdateReferencesToSecret to store referrence records between the object and referred secrets,
+	// and also to remove the outdated reference records in cache where the secret is not referred by the obj specification anymore.
+	case *corev1.Service:
+		return updateCoreV1ServiceReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *netv1.Ingress:
+		return updateNetV1IngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *netv1beta1.Ingress:
+		return updateNetV1beta1IngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *extv1beta1.Ingress:
+		return updateExtensionV1beta1IngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1.KongPlugin:
+		return updateKongPluginReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1.KongClusterPlugin:
+		return updateKongClusterPluginReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1.KongConsumer:
+		return updateKongConsumerReferredSecrets(ctx, client, dataplaneClient, obj)
+	case *kongv1beta1.TCPIngress:
+		return updateTCPIngressReferredSecrets(ctx, client, dataplaneClient, obj)
+	}
```

Comment:
May I suggest a small refactor here? Instead of calling `reference.UpdateReferencesToSecret()` in every single one of those functions, let's return `[]types.NamespacedName` from all the `update***ReferredSecrets` and call `reference.UpdateReferencesToSecret()` just once in here.

This will reduce the copy pasta and make it more DRY.

WDYT?

Reply:
I think it is a very good suggestion, will do.

Reply:
updated.